docs/how-packages-and-artifacts-are-built.md

```diff
@@ -0,0 +1,114 @@
+
+#### Sources
+Sources are singular artifacts or git repositories that are required to complete the package build. They are described via JSON maps with variable fields; the only guaranteed field is `kind`. There are a few kinds of sources:
+* `git`: Points to a git repository which can be private if the build host git client is configured correctly. Required fields: `kind`, `url`, `ref`, `ref_origin`. `ref` is the commit from the repo that should be built and `ref_origin` is the branch on which that commit live.
+* `git_local`: A filesystem-local git repository that may have been pulled in by a process outside gen/release. Required fields: `kind` and `rel_path`
+* `url`: URL of a singular artifact to download and volume mount inside the build container. Required fields: `kind`, `url`, `sha1`. `sha1` is the SHA-1 of the artifact and can be found by downloading the artifact and running `sha1sum <<my_artifact>>`
+* `url_extract`: URL is to an artifact that is a compressed file that will be automatically decompressed before volume mounting. Supported compression types are: `.tar.gz`, `.tgz`, and `.zip`. Required fields are: `kind`, `url`, and `sha1`
+
+As for describing in relation to the `buildinfo.json`:
+* If there are no required dependencies outside of the package directory, then no source information is required `buildinfo.json`
+* If there is only one source, a `source` field must be included in the `buildinfo.json`. The data from this source will be mounted inside the container as `/pkg/src/$PKG_NAME`
+* If there are multiple sources, a `sources` field must be provided in `buildinfo.json` and each of its subfields will be a source like the single source entry. Each of these sources will be mounted in the container as `/pkg/src/<<key name of source in JSON sources>>`
+
+
+#### Customized Docker Build Images
+By default, all packages are built with the [dcos-builder Dockerfile](../pkgpanda/docker/dcos-builder), however, packages can be built with any docker image. There are two ways to specify other docker images:
+* include a folder called `docker` along side the `build` and `buildinfo.json`. This file must contain a Dockerfile and any necessary data to be included with the image.
+* reference a public dockerhub image in the `buildinfo.json`
+
+#### Magic Variables in Build Containers
+* `/pkg/src/`: directory for all sources declared in `buildinfo.json`
+* `/pkg/extra/`: mounted `extra` directory from the package folder
+* `/opt/mesosphere/environment.export`: file including the environment variables that will link the build container environment to built packages and their artifacts. Most builds depending on other packages should source the environment with this file
+* `/opt/mesosphere`: the various accumulation of dependent artifacts as they will be on the host at runtime
```

Comment:
its a little more complicated than that because it also includes the symlinked superposition of lib and bin, but I think I can communicate that better...